src/renderers/shared/fiber/ReactFiberReconciler.js

```diff
@@ -54,16 +49,34 @@ module.exports = function<T, P, I>(config : HostConfig<T, P, I>) : Reconciler {
 
   let nextUnitOfWork : ?Fiber = null;
 
-  function performUnitOfWork(unit : Fiber) : ?Fiber {
-    switch (unit.tag) {
-      case FunctionalComponent:
-        return ReactFiberFunctionalComponent.performWork(unit);
-      case ClassComponent:
-        break;
-      case HostComponent:
-        break;
+  function completeUnitOfWork(unitOfWork : Fiber) : ?Fiber {
+    while (true) {
+      var next = completeWork(unitOfWork);
+      if (next) {
+        // If completing this work spawned new work, do that next.
+        return next;
+      } else if (unitOfWork.sibling) {
+        // If there is more work to do in this parent, do that next.
+        return unitOfWork.sibling;
+      } else if (unitOfWork.parent) {
+        // If there's no more work in this parent. Complete the parent.
+        unitOfWork = unitOfWork.parent;
+      } else {
+        // If we're at the root, there's no more work to do.
+        return null;
+      }
+    }
+  }
+
+  function performUnitOfWork(unitOfWork : Fiber) : ?Fiber {
+    var next = beginWork(unitOfWork);
+    if (next) {
+      // If this spawns new work, do that next.
+      return next;
```

Comment:
Completion is currently synchronous which means that the fiber doesn't need to be in two states. I guess that's an important part of this.

Completion is expected to be fast. The only user code fired would be a coroutine handler. The handler would then return another fiber so it would be at most one coroutine handler. The worst case scenario would be if all the coroutine handlers returned null. In that case we might execute many of them synchronously.